g2p/cli.py

```diff
@@ -532,38 +532,50 @@ def convert(  # noqa: C901
             f"Path between '{in_lang}' and '{out_lang}' does not exist"
         )
     # Figure if the input text is on the command line or in a file
-    input_text_is_a_file = os.path.exists(input_text) and input_text.endswith("txt")
-    if input_text_is_a_file:
-        with open(input_text, encoding="utf8") as f:
-            lines = f.readlines()
-    else:
-        lines = [input_text]
-    # Determine which tokenizer to use, if any
-    if tok is not None and not tok and tok_lang is not None:
-        raise click.UsageError("Specified conflicting --no-tok and --tok-lang options.")
-    if tok is None:
-        tok = True  # Tokenize by default
-    custom_tokenizer = make_tokenizer(tok_lang) if tok_lang else None
-    # Transduce!!!
-    assert in_lang and out_lang
-    transducer = make_g2p(
-        in_lang, out_lang, tokenize=tok, custom_tokenizer=custom_tokenizer
+    input_text_is_a_file = (
+        os.path.exists(input_text)
+        and input_text.endswith("txt")
+        or re.match(r"/dev/(fd/[0-9]*|stdin)", input_text)
     )
-    for line in lines:
-        tg = transducer(line)
-        if check:
-            transducer.check(tg, display_warnings=True)
-        outputs = [tg.output_string]
-        if substring_alignments:
-            outputs += [tg.substring_alignments()]
-        if pretty_edges:
-            outputs += [tg.pretty_edges()]
-        if debugger:
-            outputs += [tg.edges, tg.debugger]
-        if len(outputs) > 1:
-            click.echo(pprint.pformat(outputs, indent=4))
+    to_close = None
+    try:
+        if input_text_is_a_file:
```

Comment:
Can we just use click.File for this? https://click.palletsprojects.com/en/8.1.x/arguments/#file-arguments

Reply:
I don't know if click supports "string OR file" which is what we have here. The original was just a string, then for convenience if the string happened to name a file I'd open and read it, now there are users of this at-first-hidden feature and I'm trying to improve it.
click.File would be a good choice if we weren't abusing the syntax like that.
I suppose this mean suggest changing the interface to have --input-file and --input-text instead of conflating the two, but I'm not keen on making such a breaking change without a strong compelling case (stronger than reusing library code).

Reply:
oh... "string or file"? so, if it is a string that doesn't name an existing file then we treat it as a sring, otherwise we open it as a file?  This is very strange behaviour for a CLI :)

Reply:
Indeed, I rather regret that weird CLI choice indeed, especially with what you see here, but now that some people rely on it, I'm not that keen on changing it. Although it's only a feature you'd use interactively, so it wouldn't be that bad to break it: it would not be a breaking API or breaking library change, so no dependent code would get broken by it, only the people using the feature would have to learn the modified behaviour.

And "string or file" is a little narrower than that: it has to name a file that has a .txt extension and exists, which was initially just a hidden feature for my private use that some users at UBC stumbled upon because it was the obvious thing to do and it happened to work as they expected.

The annoying thing now is that this "string or file" is a mandatory positional argument, and if I were to replace the file variant with `--in FILENAME` or something like, then suddenly the first positional argument would be the optional one, and I really don't know how to implement that in click.

If we change the interface for accepting a filename, I would issue a warning when the old automatic logic would have been triggered documenting the new way to call it.

But anyway, my question now is, do you have a good suggestion for a better CLI behaviour that would be intuitive, easy to implement, and most importantly not change the original behaviour of `g2p convert INPUT_TEXT IN_LANG OUT_LANG` which has been around a long time and would, in my opinion, be a major breaking change if we modified it?

I suppose we could add a `--text-is-filename|--file|-f` Boolean flag, but I kinda prefer the current weird auto-detect behaviour... However with such a flag, we'd still have "string or file", just with simpler triggering logic.